pytorch_memlab/line_profiler.py

```diff
@@ -1,26 +1,171 @@
+
+    if columns is not None: 
+        columns = [tuple(c.split('.')) for c in columns]
+        if not all(len(c) == 3 for c in columns):
+            raise ValueError('Each column name should have three dot-separated parts')
+        if not all(c in line_records.columns for c in columns):
+            options = ", ".join(".".join(c) for c in line_records.columns.tolist())
+            raise ValueError('The column names should be fields of torch.cuda.memory_stat(). Options are: ' + options)
+        line_records = line_records.loc[:, columns]
+    
+    return line_records
+
+
+class RecordsDisplay:
+    """IPython's rich display functionality [requires we return](https://ipython.readthedocs.io/en/stable/config/integrating.html)
+    an object that has a `_repr_html_` method for when HTML rendering is supported, and 
+    a `__repr__` method for when only text is available"""
+
+    def __init__(self, line_records, code_infos):
+        if len(line_records) > 0:
+            self._line_records = line_records.groupby(level=[0, 1]).max() 
+        else:
+            self._line_records = line_records
+        self._code_infos = code_infos
+
+    def _line_records_merged_with_code(self):
+        merged = {}
+        for _, info in self._code_infos.items():
+            qual_name = info['func'].__qualname__
+            if qual_name in self._line_records.index.get_level_values(0):
+                lines, start_line = inspect.getsourcelines(info['func'])
+                lines = pd.DataFrame.from_dict({
+                    'line': range(start_line, start_line + len(lines)), 
+                    'code': lines})
+                lines.columns = pd.MultiIndex.from_product([lines.columns, [''], ['']])
+                
+                merged[qual_name] = pd.merge(
+                    self._line_records.loc[qual_name], lines, 
+                    right_on='line', left_index=True, how='right')
+        return merged
+
+    def __repr__(self):
+        """Renders the stats as text"""
+        if len(self._line_records) == 0:
+            return 'No data collected'
+
+        is_byte_col = self._line_records.columns.get_level_values(0).str.contains('byte')
+        byte_cols = self._line_records.columns[is_byte_col]
+
+        string = {}
+        for qual_name, merged in self._line_records_merged_with_code().items():
+            left_align = '{{:{maxlen}s}}'.format(maxlen=max(map(len, merged.code)))
+            merged[byte_cols] = merged[byte_cols].applymap(readable_size)
+
+            # This is a mess, but I can't find any other way to left-align text strings.
+            code_header = (left_align.format('code'), '', '')
+            merged[code_header] = merged['code'].apply(lambda l: left_align.format(l.rstrip('\n\r')))
+            merged = merged.drop('code', 1, level=0)
+
+            string[qual_name] = merged.to_string(index=False)
+
+        return '\n\n\n'.join(['## {q}\n\n{c}'.format(q=q, c=c) for q, c in string.items()])
```

Comment:
Better to move a `\n` from `'\n\n\n` after `{c}` to make a line break.

Reply:
? Doing that would leave the text with an empty line at the bottom of the output.